src/with_const_generics.rs

```diff
@@ -241,16 +241,283 @@ impl<'a, T, const CAP: usize> IntoIterator for &'a mut ConstGenericRingBuffer<T,
+                    for b in batch.iter().take(how_full) {
+                        // Safety: fill_batch filled up at least `how_full` items so if we iterate
+                        // until there this is safe
+                        self.push(unsafe { b.assume_init_read() });
+                    }
+
+                    // then we're done!
+                    return;
+                }
+
+                // else the batch is full, and we can transmute it to an init slice
+                let batch = unsafe {
+                    mem::transmute_copy::<[MaybeUninit<T>; BATCH_SIZE], [T; BATCH_SIZE]>(&batch)
+                };
+
+                // SAFETY: if CAP < Self::BATCH_SIZE we woudn't be here
+                unsafe { self.extend_from_arr_batch(batch) }
+            }
+        }
+    }
+
+    /// # Safety
+    /// ONLY USE WHEN WORKING ON A CLEARED RINGBUFFER
+    #[cfg(feature = "batched_extend")]
+    #[inline]
+    unsafe fn finish_iter<const BATCH_SIZE: usize>(&mut self, mut iter: impl Iterator<Item = T>) {
+        let mut index = 0;
+        for i in iter.by_ref() {
+            self.buf[index] = MaybeUninit::new(i);
+            index += 1;
+
+            if index > CAP - 1 {
+                break;
+            }
+        }
+
+        if index < CAP {
+            // we set writepointer to however many elements we managed to write (up to CAP-1)
+            // WARNING: ONLY WORKS WHEN WORKING ON A CLEARED RINGBUFFER
+            self.writeptr = index;
+        } else {
+            self.writeptr = CAP;
+            self.extend_batched_internal::<BATCH_SIZE>(iter);
+        }
+    }
+
+    /// Alias of [`Extend::extend`](ConstGenericRingBuffer::extend) but can take a custom batch size.
+    ///
+    /// We found that `30` works well for us, which is the batch size we use in `extend`,
+    /// but on different architectures this may not be true.
+    pub fn custom_extend_batched<const BATCH_SIZE: usize>(
+        &mut self,
+        iter: impl IntoIterator<Item = T>,
+    ) {
+        #[cfg(not(feature = "batched_extend"))]
+        {
+            for i in iter {
+                self.push(i);
+            }
+        }
+
+        #[cfg(feature = "batched_extend")]
+        {
+            let iter = iter.into_iter();
+
+            let (lower, _) = iter.size_hint();
+
+            if lower >= CAP {
+                // if there are more elements in our iterator than we have size in the ringbuffer
+                // drain the ringbuffer
+                self.clear();
+
+                // we need exactly CAP elements.
+                // so we need to drop until the number of elements in the iterator is exactly CAP
+                let num_we_can_drop = lower - CAP;
+
+                let iter = iter.skip(num_we_can_drop);
+
+                // Safety: clear above
+                unsafe { self.finish_iter::<BATCH_SIZE>(iter) };
+            } else if self.is_empty() {
+                self.clear();
```

Comment:
I think you can move this into `finish_iter` and make it safe. Maybe rename it to `new_from_iter` or something.